python/ray/tune/tuner.py

```diff
@@ -0,0 +1,176 @@
+
+    Args:
+        trainable (Union[
+                str,
+                Callable,
+                Type[Trainable],
+                Type[ConvertibleToTrainable],
+                ConvertibleToTrainable,
+            ]): The trainable to be tune.
+        param_space (Optional[Dict[str, Any]]): Search space to run tuning on.
+            One thing to note is that both preprocessor and dataset can be tuned here.
+        tune_algo_config (Optional[TuneAlgoConfig]): tune algo specific configs.
+            Refer to ray.tune.tuner_config.TunerAlgoConfig for more info.
+        name (Optional[str]): Name of the run.
+        local_dir (Optional[str]): Local dir to save training results to.
+            Defaults to ``~/ray_results``.
+        callbacks (Optional[List[Callback]]): Callbacks to invoke.
+            Refer to ray.tune.callback.Callback for more info.
+
+    Usage pattern:
+    .. code-block:: python
+
+    # Only do the following, if you want to run in ray client mode.
+    # This means the tune driver code will be running on head node of your cluster.
+    ray.init("ray://my_ray_cluster_address")
+
+    param_space = {
+        "scaling_config": {
+            "num_actors": tune.grid_search([2, 4]),
+            "cpus_per_actor": 2,
+            "gpus_per_actor": 0,
+        },
+        "preprocessor": tune.grid_search([prep_v1, prep_v2]),
+        "datasets": {
+            "train_dataset": tune.grid_search([ds1, ds2]),
+        },
+        "params": {
+            "objective": "binary:logistic",
+            "tree_method": "approx",
+            "eval_metric": ["logloss", "error"],
+            "eta": tune.loguniform(1e-4, 1e-1),
+            "subsample": tune.uniform(0.5, 1.0),
+            "max_depth": tune.randint(1, 9),
+        },
+    }
+    tuner = Tuner(trainable=trainer, param_space=param_space, name="my_tune_run")
+    analysis = tuner.fit()
+
+    This returns an ``ExperimentAnalysis`` object, that you can interact with according
+    to its API.
+
+    If the run finishes in error, to retry
+    you can then do
+    .. code-block:: python
+
+    tuner = Tuner.restore(experiment_checkpoint_dir)
+    tuner.fit()
+
+    to resume your run.
+    """
+
+    # One of the following is assigned.
+    _local_tuner: Optional[TunerInternal]  # Only used in none ray client mode.
+    _remote_tuner: Optional[ClientActorHandle]  # Only used in ray client mode.
+
+    def __init__(
+        self,
+        trainable: Optional[
+            Union[
+                str,
+                Callable,
+                Type[Trainable],
+                Type[ConvertibleToTrainable],
+                ConvertibleToTrainable,
+            ]
+        ] = None,
+        param_space: Optional[Dict[str, Any]] = None,
+        tune_algo_config: Optional[TuneAlgoConfig] = None,
+        name: Optional[str] = None,
+        local_dir: Optional[str] = None,
+        callbacks: Optional[List[Callback]] = None,
```

Comment:
Are these args supposed to be grouped together as the RunConfig?

Reply:
+1

Reply:
Addressed in the API review.